oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/UnrecoveredRevisionTest.java

```diff
@@ -0,0 +1,813 @@
+     */
+    @Test
+    public void testAddNode_variantB() throws Exception {
+        doTestAddNode(1, "/foo/a/b");
+    }
+
+    private void doTestRemoveNode(int createOrUpdateBatchSize, String targetTestNode) throws Exception {
+        setup(createOrUpdateBatchSize);
+        
+        // test setup, using ns1
+        LeavesBuilder leaves = 
+                new LeavesBuilder().add("/foo").add("/foo/a/b/c/d", "/foo/e/f/g/h", "/foo/e/f/i/j");
+        createContent(leaves, ns1, ns2);
+
+        // now produce JVM-pause/lateWrite/crash situation, using ns2
+        produceLateWriteSituation(2, ns2, (rb) -> removeNodes(rb, "/foo/a/b/c/d", "/foo/e/f/g/h"));
+
+        // ensure the above didn't manage to delete properly
+        advanceClockOneSecond();
+        leaves.assertLeavesExist(ns1);
+        
+        // now cause a conflict due to the late write
+        try {
+            setProperty(2, targetTestNode, "p2", "v2");
+            fail("should fail");
+        } catch(CommitFailedException cfe) {
+            assertTrue(cfe.toString().contains("OakMerge0004"));
+        }
+    }
+    
+    /**
+     * Illustrates a late-write remove-node can later on cause an OakMerge0004 conflict
+     */
+    @Test
+    public void testRemoveNode_variantA() throws Exception {
+        doTestRemoveNode(1, "/foo/a/b/c/d");
+    }
+
+    /**
+     * Illustrates a late-write remove-node can later on cause an OakMerge0004 conflict
+     */
+    @Test
+    public void testRemoveNode_variantB() throws Exception {
+        doTestRemoveNode(1, "/foo/e/f/g/h");
+    }
+
+    /**
+     * Illustrates a late-write remove-node can later on cause an OakMerge0004 conflict
+     */
+    @Test
+    public void testRemoveNode_variantC() throws Exception {
+        doTestRemoveNode(10, "/foo/e/f/g/h");
+    }
+    
+    private void runBackgroundOps(int times, DocumentNodeStore... dnss) {
+        for( int i = 0; i < times; i++) {
+            for (DocumentNodeStore dns : dnss) {
+                dns.runBackgroundOperations();
+            }
+        }
+    }
+    
+    private void advanceClock(long duration, TimeUnit unit) throws InterruptedException {
+        advanceClock(unit.toMillis(duration));
+    }
+
+    private void advanceClockOneSecond() throws InterruptedException {
+        advanceClock(1, TimeUnit.SECONDS);
+    }
+    
+    private void printNodes(DocumentNodeStore ns, String... paths) {
+        for (String aPath : paths) {
+            Path p = Path.fromString(aPath);
+            NodeState n = ns.getRoot();
+            for (String name : p.elements()) {
+                n = n.getChildNode(name);
+            }
+            System.out.println(aPath + " :");
+            if (!n.exists()) {
+                System.out.println("DOES NOT EXIST");
+            } else {
+                DocumentNodeState dns = (DocumentNodeState) n;
+                NodeDocument doc = dns.getDocument();
```

Comment:
This method does not exist. Rather get the path from the DocumentNodeState and then look up the document from the DocumentStore?

Reply:
oups, missed to clean that up, had that locally hacked in.

now fixed in https://github.com/apache/jackrabbit-oak/pull/950/commits/3846d96be9eb30904a1bc836c8e968525f579faf